### ISSUE TITLE
Timeout fails to stop test

### ISSUE THREAD

**Issue by saiema** [junit-team/junit5#1742]
Hello, I didn't find this issue reported.

On JUnit 5 (5.3.1.v20181005-1442) and also in JUnit 4 (4.12.0.v201504281640), when running the test below on eclipse (Version: 2018-12 (4.10.0) Build id: 20181214-0600) on Ubuntu 18.04, the timeout rule fails to stop de test.

    @Rule
    public Timeout globalTimeout = Timeout.millis(1000);

    @Test
    void test() {
	NaughtyList list = new NaughtyList();
        list.add(1);
    }

For the following class

    public class NaughtyList {
	
	public static class NaughtyNode {
		public int value;
		public NaughtyNode next;
	}
	
	private NaughtyNode header;
	
	public NaughtyList() {
		header = new NaughtyNode();
		header.value = Integer.MIN_VALUE;
	}
	
	public void add(int elem) {
		NaughtyNode current = header;
		while(true) {
			current.next = new NaughtyNode();
			current = current.next;
			current.value = elem;
		}
	}

    }

I know that this is an extreme case but I found this issue when doing some mutation testing experiments and some mutants caused an infinite loop on a very simple code that kept creating new objets.

**Comment by marcphilipp** [comment]
JUnit calls `Thread.interrupt()` on the timed out thread. In your case there's no blocking operation that would throw an `InterruptedException` and the code under test does not check for the thread's interrupted flag (`Thread.interrupted()`), there's nothing we can do to stop it.

**Comment by sbrannen** [comment]
FWIW, the test actually does fail for me with JUnit 4 (running on JDK 11):

```
org.junit.runners.model.TestTimedOutException: test timed out after 1000 milliseconds
	at app//foo.NaughtyList.add(NaughtyList.java:21)
	at app//foo.NaughtyListTests.test(NaughtyListTests.java:16)
	at java.base@11.0.1/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.1/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.1/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.1/java.lang.reflect.Method.invoke(Method.java:566)
	at app//org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at app//org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at app//org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at app//org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at app//org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at app//org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.base@11.0.1/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base@11.0.1/java.lang.Thread.run(Thread.java:834)

```

**Comment by sbrannen** [comment]
And when converted to JUnit Jupiter (JUnit 5)...

```java
import static java.time.Duration.ofMillis;
import static org.junit.jupiter.api.Assertions.assertTimeoutPreemptively;

import org.junit.jupiter.api.Test;

class NaughtyListTests {

	@Test
	void test() {
		NaughtyList list = new NaughtyList();
		assertTimeoutPreemptively(ofMillis(1000), () -> list.add(1));
	}

}
```

... the test also fails as follows:

```
org.opentest4j.AssertionFailedError: execution timed out after 1000 ms
	at org.junit.jupiter.api.AssertTimeout.assertTimeoutPreemptively(AssertTimeout.java:144)
	at org.junit.jupiter.api.AssertTimeout.assertTimeoutPreemptively(AssertTimeout.java:115)
	at org.junit.jupiter.api.AssertTimeout.assertTimeoutPreemptively(AssertTimeout.java:97)
	at org.junit.jupiter.api.AssertTimeout.assertTimeoutPreemptively(AssertTimeout.java:93)
	at org.junit.jupiter.api.Assertions.assertTimeoutPreemptively(Assertions.java:2743)
	at foo.NaughtyListTests.test(NaughtyListTests.java:14)
...
```

**Comment by marcphilipp** [comment]
Well, the test fails, but the rogue thread never stops. I assumed that's what the original bug report was about.

**Comment by sbrannen** [comment]
> Well, the test fails, but the rogue thread never stops. I assumed that's what the original bug report was about.

Ahaaaaaaaaaaaaaa... OK. Sorry for the mix-up.

Test test _fails_, but the user code may potentially _execute indefinitely_. 😉

**Comment by sbrannen** [comment]
I renamed this issue accordingly.

**Comment by saiema** [comment]
Hi.

I tested the following (I'm using Java 8, Oracle JDK):

In the add method, if I just leave the while(true) without any body I get the following:

Using JUnit 5 with the timeout Rule: the test never ends (it never fails)
Using JUnit 4 either with timeout annotation or a timeout Rule: the test fails after 1 second 

> Well, the test fails, but the rogue thread never stops. I assumed that's what the original bug report was about.

On previous experiments I got that issue when running a mutation analysis and the solution was an ugly hack that after the analysis it will find and stop any JUnit threads. But currently when a simple code that runs on an endless loop and creates new objects in the code, the test does not finish

**Comment by saiema** [comment]
> FWIW, the test actually does fail for me with JUnit 4 (running on JDK 11):
> 
> ```
> org.junit.runners.model.TestTimedOutException: test timed out after 1000 milliseconds
> 	at app//foo.NaughtyList.add(NaughtyList.java:21)
> 	at app//foo.NaughtyListTests.test(NaughtyListTests.java:16)
> 	at java.base@11.0.1/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base@11.0.1/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base@11.0.1/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base@11.0.1/java.lang.reflect.Method.invoke(Method.java:566)
> 	at app//org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> 	at app//org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at app//org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> 	at app//org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at app//org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
> 	at app//org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
> 	at java.base@11.0.1/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> 	at java.base@11.0.1/java.lang.Thread.run(Thread.java:834)
> ```

That's interesting, I can only use Java 7 or 8 to run my experiments, maybe it's a garbage collector issue?

**Comment by saiema** [comment]
For some strange reason the test (including the original add code) fails after one second. Maybe a rogue thread was the cause. I will run a few experiments to see what is happening, I was having issues for almost three days.

**Comment by saiema** [comment]
> > Using JUnit 5 with the timeout Rule: the test never ends (it never fails)
> 
> JUnit Jupiter does not support JUnit 4' timeout Rule. So you can never expect that to work (i.e., fail the test).
> 
> With Jupiter you need to use `assertTimeoutPreemptively(...)` as I demonstrated above.
> 
> Keep in mind that we plan to add support for global timeouts in a later JUnit Jupiter release. See #80 for details.

Ok, I can't use JUnit 5 or newer on my tool, since I use custom runners that depend on inner code of JUnit 4, changing that code to use a newer version would require too much work, I worst case scenario I will drop the case for which I have this issue.

**Comment by saiema** [comment]
Ok, I read about how JUnit runs timedout tests and how you can (without using ugly code) stop or pause threads. It seems that when JUnit states that a test failed because of a timeout, the Thread running the users code doesn't necessarily stops (I forgot about that). I tried to see if there was a way to at least reduce the priority of the users thread once a test was marked as timedout but Thread#setPriority(int) doesn't seem to work.

The only way to tackle this issue would be to try and separate user's tests on several processes so once each process is finish the created threads are destroyed, but this would impose a significant overhead.

Sorry for the misunderstanding,  seeing the "test failed" in JUnit made me think that the user's code was being terminated.

**Comment by sbrannen** [comment]
> Ok, I can't use JUnit 5 or newer on my tool, since I use custom runners that depend on inner code of JUnit 4, changing that code to use a newer version would require too much work, I worst case scenario I will drop the case for which I have this issue.

OK. Thanks for the feedback.

**Comment by Vazzilena** [comment]
> > Using JUnit 5 with the timeout Rule: the test never ends (it never fails)
> 
> JUnit Jupiter does not support JUnit 4' timeout Rule. So you can never expect that to work (i.e., fail the test).
> 
> With Jupiter you need to use `assertTimeoutPreemptively(...)` as I demonstrated above.
> 
> Keep in mind that we plan to add support for global timeouts in a later JUnit Jupiter release. See #80 for details.


Guys, 

We started experiencing a very similar issue when we started migrating to junit5 runners and switched to junit-platform-console-standalone-1.6.0-RC1.jar

We are running our junit4 tests with the vintage engine and I am kind of confused – isn’t vintage supposed to support junit4 and junit3 tests?

Is there a way to workaround this without having to migrate all the tests to junit5? 

What we observe: 

JUnit4's FailOnTimeout rule would create a separate thread group and thread to run the test in (including any other rule's before()/after()) : https://github.com/junit-team/junit4/blob/master/src/main/java/org/junit/internal/runners/statements/FailOnTimeout.java 

JUnit5 does it differently, and it would try to run rule's after() statements on its own (in its own main thread). This results in rule's before() and after() statements run in separate threads and run in parallel which must not happen and eventually causes the `IllegalThreadStateException `in our tests.
